pkg/controller/metrics/metrics_with_dynamic_labels.go

```diff
@@ -0,0 +1,177 @@
+package metrics
+
+import (
+	"github.com/prometheus/client_golang/prometheus"
+	log "github.com/sirupsen/logrus"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/metrics"
+
+	hivev1 "github.com/openshift/hive/apis/hive/v1"
+	"github.com/openshift/hive/apis/hive/v1/metricsconfig"
+	"github.com/openshift/hive/pkg/constants"
+)
+
+// metricsWithDynamicLabels encompasses metrics that can have optional labels.
+type metricsWithDynamicLabels interface {
+	ObserveMetricWithDynamicLabels(metav1.Object, log.FieldLogger, map[string]string, float64)
+	RegisterMetricWithDynamicLabels()
+}
+
+type dynamicLabels struct {
+	// fixedLabels defines the label keys that are always registered on the metric.
+	fixedLabels []string
+	// optionalLabels is generated dynamically based on HiveConfig.Spec.MetricsConfig.MetricsWithDynamicLabels, and
+	// represents a map of metric label keys to ClusterDeployment metadata.label keys from which values will be gleaned
+	// when the metric is observed.
+	optionalLabels map[string]string
+}
+
+func WithFixedLabels(labels []string) func(*dynamicLabels) {
+	return func(d *dynamicLabels) {
+		d.fixedLabels = labels
+	}
+}
+
+func WithOptionalLabels(labels map[string]string) func(*dynamicLabels) {
+	return func(d *dynamicLabels) {
+		d.optionalLabels = labels
+	}
+}
+
+// getLabelList combines fixed and optional labels and returns the list without duplicates.
```

Comment:
While doing this, I thought about how it would be easier to have an early feedback - if a user tries to cause an overlap between these labels. However, there is not simpler way to do it, and we wouldn't want to error out because of metrics.